File: src-tauri/src/commands/speech.rs
use crate::domain::speech::speech_error;
use crate::domain::{
    speech::{SpeechModel, SpeechModelStatus, SpeechProgress, SpeechTranscript},
    AiSettings, AppError,
};
use crate::services::speech::SpeechService;
use std::path::PathBuf;
use tauri::{AppHandle, Emitter, State};

#[tauri::command]
pub fn speech_model_status(
    service: State<'_, SpeechService>,
    model: Option<SpeechModel>,
) -> SpeechModelStatus {
    service.status(model.unwrap_or_default())
}

#[tauri::command]
pub async fn install_speech_model(
    app: AppHandle,
    service: State<'_, SpeechService>,
    request_id: String,
    source: Option<String>,
    model: Option<SpeechModel>,
) -> Result<SpeechModelStatus, AppError> {
    let service = service.inner().clone();
    tauri::async_runtime::spawn_blocking(move || {
        service.install_model(
            &request_id.clone(),
            source.map(PathBuf::from),
            model.unwrap_or_default(),
            move |progress| {
                let _ = app.emit(
                    "speech-progress",
                    SpeechProgress {
                        request_id: request_id.clone(),
                        stage: "downloading",
                        progress,
                    },
                );
            },
        )
    })
    .await
    .map_err(|_| speech_error("downloadError"))?
}

#[tauri::command]
pub async fn start_speech_recording(
    service: State<'_, SpeechService>,
    request_id: String,
    model: Option<SpeechModel>,
) -> Result<(), AppError> {
    let service = service.inner().clone();
    tauri::async_runtime::spawn_blocking(move || {
        service.start(&request_id, model.unwrap_or_default())
    })
    .await
    .map_err(|_| speech_error("microphoneError"))?
}

#[tauri::command]
pub async fn stop_speech_recording(
    app: AppHandle,
    service: State<'_, SpeechService>,
    request_id: String,
    language: String,
) -> Result<SpeechTranscript, AppError> {
    let service = service.inner().clone();
    tauri::async_runtime::spawn_blocking(move || {
        service.stop(&request_id.clone(), language, move |progress| {
            let _ = app.emit(
                "speech-progress",
                SpeechProgress {
                    request_id: request_id.clone(),
                    stage: "transcribing",
                    progress,
                },
            );
        })
    })
    .await
    .map_err(|_| speech_error("transcribeError"))?
}

#[tauri::command]
pub fn cancel_speech(service: State<'_, SpeechService>, request_id: String) {
    service.cancel(&request_id);
}

#[tauri::command]
pub async fn format_speech_transcript(
    settings: AiSettings,
    text: String,
) -> Result<String, AppError> {
    tauri::async_runtime::spawn_blocking(move || {
        crate::infrastructure::ai::ChatCompletionClient::new(&settings)?.format_transcript(&text)
    })
    .await
    .map_err(|_| speech_error("formatError"))?
}

File: src/gateways/speech.ts
import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";
import { open } from "@tauri-apps/plugin-dialog";
import type { SpeechGateway } from "./contracts";
import type { SpeechModelStatus, SpeechProgress, SpeechTranscript } from "../domain/speech";
import { GatewayError, mapGatewayError } from "../domain/errors";

async function call<T>(command: string, args?: Record<string, unknown>): Promise<T> {
  try { return await invoke<T>(command, args); } catch (error) { throw mapGatewayError(error); }
}

export function createTauriSpeechGateway(): SpeechGateway {
  return {
    available: true,
    modelStatus: (model = "small") => call<SpeechModelStatus>("speech_model_status", { model }),
    chooseModel: async (model = "small") => {
      const path = await open({ multiple: false, filters: [{ name: `Whisper ${model} Q5_1`, extensions: ["bin"] }] });
      return typeof path === "string" ? path : null;
    },
    installModel: (requestId, source, model = "small") => call<SpeechModelStatus>("install_speech_model", { requestId, source, model }),
    start: (requestId, model = "small") => call<void>("start_speech_recording", { requestId, model }),
    stop: (requestId, language, context) => call<SpeechTranscript>("stop_speech_recording", { requestId, language, context }),
    cancel: (requestId) => call<void>("cancel_speech", { requestId }),
    format: (settings, text) => call<string>("format_speech_transcript", { settings, text }),
    watchProgress: (onProgress) => listen<SpeechProgress>("speech-progress", (event) => onProgress(event.payload)),
  };
}

export function createUnavailableSpeechGateway(): SpeechGateway {
  const unavailable = async (): Promise<never> => { throw new GatewayError({ code: "io", message: "speech.desktopOnly" }); };
  return {
    available: false, modelStatus: unavailable, chooseModel: unavailable, installModel: unavailable,
    start: unavailable, stop: unavailable, format: unavailable,
    cancel: async () => {}, watchProgress: async () => () => {},
  };
}
